packages/andromeda-non-fungible-tokens/src/crowdfund.rs

```diff
@@ -1,111 +1,162 @@
+    READY,
+    /// Stage when campaign is being carried out
+    ONGOING,
+    /// Stage when campaign is finished successfully
+    SUCCESS,
+    /// Stage when campaign failed to meet the target cap before expiration
+    FAILED,
+}
+
+impl ToString for CampaignStage {
+    #[inline]
+    fn to_string(&self) -> String {
+        match self {
+            Self::READY => "READY".to_string(),
+            Self::ONGOING => "ONGOING".to_string(),
+            Self::SUCCESS => "SUCCESS".to_string(),
+            Self::FAILED => "FAILED".to_string(),
+        }
+    }
 }
 
 #[cw_serde]
-pub struct Config {
-    /// The address of the token contract whose tokens are being sold.
-    pub token_address: AndrAddr,
-    /// Whether or not the owner can mint additional tokens after the sale has been conducted.
-    pub can_mint_after_sale: bool,
+pub struct Tier {
+    pub level: Uint64,
+    pub label: String,
+    pub price: Uint128,
+    pub limit: Option<Uint128>, // None for no limit
+    pub meta_data: TierMetaData,
 }
 
 #[cw_serde]
-pub struct State {
-    /// The expiration denoting when the sale ends.
-    pub end_time: Expiration,
-    /// The price of each token.
-    pub price: Coin,
-    /// The minimum number of tokens sold for the sale to go through.
-    pub min_tokens_sold: Uint128,
-    /// The max number of tokens allowed per wallet.
-    pub max_amount_per_wallet: u32,
-    /// Number of tokens sold.
-    pub amount_sold: Uint128,
-    /// The amount of funds to send to recipient if sale successful. This already
-    /// takes into account the royalties and taxes.
-    pub amount_to_send: Uint128,
-    /// Number of tokens transferred to purchasers if sale was successful.
-    pub amount_transferred: Uint128,
-    /// The recipient of the raised funds if the sale is successful.
-    pub recipient: Recipient,
+pub struct TierOrder {
+    pub orderer: Addr,
+    pub level: Uint64,
+    pub amount: Uint128,
 }
 
+impl Tier {
+    pub fn validate(&self) -> Result<(), ContractError> {
+        ensure!(
+            !self.price.is_zero(),
+            ContractError::InvalidTier {
+                operation: "all".to_string(),
+                msg: "Price can not be zero".to_string()
+            }
+        );
+        ensure!(
+            !self.label.is_empty() && self.label.len() <= 64,
+            ContractError::InvalidTier {
+                operation: "all".to_string(),
+                msg: "Label should be no-empty and its length can be 64 at maximum".to_string()
+            }
+        );
+
+        Ok(())
+    }
+}
 #[cw_serde]
-pub struct CrowdfundMintMsg {
-    /// Unique ID of the NFT
-    pub token_id: String,
-    /// The owner of the newly minter NFT
-    pub owner: Option<String>,
-    /// Universal resource identifier for this NFT
+pub struct TierMetaData {
+    /// Universal resource identifier for the tier
     /// Should point to a JSON file that conforms to the ERC721
```

Comment:
Shouldn't that be `CW721` instead of `ERC721` ?

Reply:
Resolved.